File: 0977-squares-of-a-sorted-array/0977-squares-of-a-sorted-array.cpp
class Solution {
public:
    vector<int> sortedSquares(vector<int>& nums) {

        // Versão NÃO Otimizada
        /*for(int i = 0; i < nums.size(); i++){
            nums[i] = pow(nums[i], 2);
        }
        sort(nums.begin(), nums.end()); //O(n log n)
        
        return nums;*/ 

        // Versão UM pouco mais Otimizada
        /* for(int i = 0; i < nums.size(); i++){
            nums[i] = nums[i] * nums[i];
        }
        sort(nums.begin(), nums.end()); //O(n log n)
        
        return nums;*/ 


        // Versão Otimizada usando ponteiros (O(n))
        int esquerda = 0;
        int direita = nums.size() - 1;
        vector<int> resultado(nums.size());

        for(int i = nums.size() - 1; i >= 0; i--){
            if(abs(nums[esquerda] * nums[esquerda]) > abs(nums[direita] * nums[direita])){
                resultado[i] = abs(nums[esquerda] * nums[esquerda]);
                esquerda++;
            } else {
                resultado[i] = abs(nums[direita] * nums[direita]);
                direita--;
            }
        }

        return resultado;
        
    }
};

File: 0977-squares-of-a-sorted-array/0977-squares-of-a-sorted-array.py
class Solution:
    def sortedSquares(self, nums: list[int]) -> list[int]:
        
        # Versão UM pouco mais Otimizada O(n \log n)

        for i in range(len(nums)):
            nums[i] = nums[i] * nums[i]
        nums.sort()
        return nums
        '''

        # Versão Otimizada usando ponteiros O(n)
        n = len(nums)
        esquerda = 0
        direita = n - 1
        
        resultado = [0] * n 


        for i in range(n - 1, -1, -1):
            quadrado_esq = nums[esquerda] * nums[esquerda]
            quadrado_dir = nums[direita] * nums[direita]
            

            if quadrado_esq > quadrado_dir:
                resultado[i] = quadrado_esq
                esquerda += 1
            else:
                resultado[i] = quadrado_dir
                direita -= 1

        return resultado
        '''
        